File: legacy/csharp/components/Strata.Application/Service/AbstractServiceEvent.cs
//  ##########################################################################
//  # File Name: AbstractServiceEvent.cs
//  # Copyright: 2018, Sapientia Systems, LLC.
//  ##########################################################################

using System;

namespace Strata.Application.Service
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// </summary>
    ///  
    public abstract
    class AbstractServiceEvent<S>:
        IServiceEvent<S>
    {
        public virtual string   Name { get; protected set; }
        public virtual DateTime Timestamp { get; protected set; }
        public virtual S        Source { get; protected set; }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        ///  
        protected
        AbstractServiceEvent(string name,S source):
            this(name,DateTime.Now,source) {}

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        ///  
        protected
        AbstractServiceEvent(string name,DateTime timestamp,S source)
        {
            Name      = name;
            Timestamp = timestamp;
            Source    = source;
        }
    }
}

//  ##########################################################################

File: java/Components/Strata.Application.Core/src/strata/application/core/service/AbstractServiceEvent.java
//////////////////////////////////////////////////////////////////////////////
// AbstractServiceEvent.java
//////////////////////////////////////////////////////////////////////////////

package strata.application.core.service;

import java.time.LocalDateTime;

/*****************************************************************************
 * Base implementation of {@code IServiceEvent<S>} interface.
 *
 * @param <S> event source type
 */
public abstract
class AbstractServiceEvent<S>
    implements IServiceEvent<S>
{
    private final String name;
    private final LocalDateTime timestamp;
    private final S             source;

    /*************************************************************************
     * Creates a new instance of {@code AbstractServiceEvent<S>}.
     *
     * @param nm  event name
     * @param src event source
     */
    protected
    AbstractServiceEvent(String nm,S src)
    {
        name = nm;
        timestamp = LocalDateTime.now();
        source = src;
    }

    /*************************************************************************
     * {@inheritDoc}
     */
    @Override
    public String
    getName()
    {
        return name;
    }

    /*************************************************************************
     * {@inheritDoc}
     */
    @Override
    public LocalDateTime
    getTimestamp()
    {
        return timestamp;
    }

    /*************************************************************************
     * {@inheritDoc}
     */
    @Override
    public S
    getSource()
    {
        return source;
    }
}

//////////////////////////////////////////////////////////////////////////////
